tianshou/trainer/offline.py

```diff
@@ -0,0 +1,95 @@
+import time
+import tqdm
+from collections import defaultdict
+from torch.utils.tensorboard import SummaryWriter
+from typing import Dict, List, Union, Callable, Optional
+
+from tianshou.policy import BasePolicy
+from tianshou.utils import tqdm_config, MovAvg
+from tianshou.data import Collector, ReplayBuffer
+from tianshou.trainer import test_episode, gather_info
+
+
+def offline_trainer(
+    policy: BasePolicy,
+    buffer: ReplayBuffer,
+    test_collector: Collector,
+    max_epoch: int,
+    step_per_epoch: int,
+    episode_per_test: Union[int, List[int]],
+    batch_size: int,
+    test_fn: Optional[Callable[[int, Optional[int]], None]] = None,
+    stop_fn: Optional[Callable[[float], bool]] = None,
+    save_fn: Optional[Callable[[BasePolicy], None]] = None,
+    writer: Optional[SummaryWriter] = None,
+    log_interval: int = 1,
+    verbose: bool = True,
+) -> Dict[str, Union[float, str]]:
+    """A wrapper for offline trainer procedure.
+
+    The "step" in trainer means a policy network update.
+
+    :param policy: an instance of the :class:`~tianshou.policy.BasePolicy`
+        class.
+    :param test_collector: the collector used for testing.
+    :type test_collector: :class:`~tianshou.data.Collector`
+    :param int max_epoch: the maximum of epochs for training. The training
+        process might be finished before reaching the ``max_epoch``.
+    :param int step_per_epoch: the number of step for updating policy network
+        in one epoch.
+    :param episode_per_test: the number of episodes for one policy evaluation.
+    :param int batch_size: the batch size of sample data, which is going to
+        feed in the policy network.
+    :param function test_fn: a function receives the current number of epoch
```

Comment:
A function defined by the user and called automatically by the algorithm is called a "hook", aka [Pytorch doc](https://pytorch.org/tutorials/beginner/former_torchies/nnft_tutorial.html#forward-and-backward-function-hooks).

:param function test_fn: a hook called at the beginning of testing in each epoch. It can be used to perform custom additional operations. It must have the following signature: f(num_peoch: int, step_idx: int) -> None.